### ISSUE TITLE
unable to build UWP app with dot net tool chain enabled

### ISSUE THREAD

**Issue by beamer15** [shimat/opencvsharp#309]
When I reference OpenCVSharp 3 and try to build with dot net tool chain enabled, I see the following error:

**Internal compiler error: An item with the same key has already been added.**		

I have created a repo that reproduces the error:

https://github.com/beamer15/ocvs3_nativetoolchain

Per this [link](https://social.msdn.microsoft.com/Forums/windowsapps/en-US/806fb3d5-0b3a-4935-b5ce-cd99f5e820bd/uwpnetnative-error-internal-compiler-error-an-item-with-the-same-key-has-already-been-added?forum=wpdevelop#fd2f3d56-1368-4de9-bf61-fffcb5cde5ab) I am also emailing the dot net native team.  I will repost with any response from them.

**Comment by MattWhilden** [comment]
Hello! I work on the .NET Native compiler team and have started to investigate this issue. The compiler seems to be having a rough time during the phase where we are trying to generate all of the C#->C++ interop code. In particular, it seems to find duplicate PInvoke signatures that it cannot reason about. I'll know more in a day or two but wanted to let you know what my cursory investigation revealed.

**Comment by beamer15** [comment]
Thanks for the update and for looking into this!

Sent from Yahoo Mail on Android 
 
  On Tue, Feb 21, 2017 at 9:25 PM, Matthew Whilden<notifications@github.com> wrote:   
Hello! I work on the .NET Native compiler team and have started to investigate this issue. The compiler seems to be having a rough time during the phase where we are trying to generate all of the C#->C++ interop code. In particular, it seems to find duplicate PInvoke signatures that it cannot reason about. I'll know more in a day or two but wanted to let you know what my cursory investigation revealed.

**Comment by MattWhilden** [comment]
So it appears that we have no way to deal with PInvoke signatures that differ only by an argument being a single or multidimensional array. In particular, here's the two that cause us to choke:

```
[DllImport("OpenCvSharpExtern", CallingConvention = CallingConvention.Cdecl)]
public static extern int core_Mat_nSetB(IntPtr obj, int row, int col, [MarshalAs(42)] [In] byte[] vals, int valsLength);
```

```
[DllImport("OpenCvSharpExtern", CallingConvention = CallingConvention.Cdecl)]
public static extern int core_Mat_nSetB(IntPtr obj, int row, int col, [MarshalAs(42)] [In] byte[,] vals, int valsLength);
```

The root cause of this is because we end up marshalling both of these as single dimensional arrays, causing the collision. I've opened a bug in our internal tracker to see what can be done in the compiler for a future release.

Presuming you don't need these PInvokes for your specific use of this library, the easiest workaround would be to close the repo and remove the "broken" multidimensional import.

I'm not certain but I believe that the "vanilla" .NET Framework/Runtime also has issues around marshalling multidimensional array so this PInvoke may not even work on non Native runtimes. Maybe?

Let me know if you have other questions/concerns. Happy to help.

**Comment by shimat** [comment]
Thank you for your information 👍 
I removed overloaded P/Invoke methods that differ only by an argument being a single or multidimensional array, and then the internal compiler error was not occurred in my environment.

Please try 3.2.0.20170224-beta-35 OpenCvSharp NuGet package in the beta package source (https://ci.appveyor.com/nuget/opencvsharp) .

```c#
[DllImport("OpenCvSharpExtern", CallingConvention = CallingConvention.Cdecl)]
public static extern int core_Mat_nSetB(IntPtr obj, int row, int col, byte* vals, int valsLength);
```
```c#
byte[] vals = new byte[1];
fixed (byte* pVals = vals)
{
    core_Mat_nSetB(obj, 0, 0, pVals, 1);
}
```
```c#
byte[,] vals = new byte[1,1];
fixed (byte* pVals = vals)
{
    core_Mat_nSetB(obj, 0, 0, pVals, 1);
}
```

**Comment by beamer15** [comment]
It's compiling, but now I'm seeing the following error on startup:

_The program XX has exited with code -1073741511 (0xc0000139) 'Entry Point Not Found'.
Activation of the Windows Store app YY failed with error 'The app didn't start'._

Any ideas?

**Comment by MattWhilden** [comment]
I suspect it's either missing a native dll or the one that is getting packaged with your app is for the wrong architecture. On non-Native runtimes this will cause crashes (Exceptions) when those dlls try to get loaded but launch issues in our runtime.

A tool like [depends.exe](http://dependencywalker.com/) can help track that down. I'm also happy to take another look if you send me a new ilcRepro file.

I feel like we're getting close! :-)

**Comment by beamer15** [comment]
Unfortunately, I was unable to glean any info from the dependency walker. The static analysis didn't show anything out of the ordinary and profiling was stifled by UWP executing apps in its container.  So, I have created an ilcRepro and uploaded at the following link.  Any help would be much appreciated!

[ilcRepro](https://1drv.ms/u/s!Ar3xk1agRxQvb93aujijjYV8qE8)

Thanks!

**Comment by MattWhilden** [comment]
Happy to help. This is the kind of error that is actually super hard to debug with .NET Native and we're going to have to put something in to help folks out. 

The issue is that one of the native dlls that we think your application needs isn't being included. In this particular case it's "OpenCvSharpExtern.dll". I'm not sure where that dll has gotten lost along the way... presumably it should be coming down via the nuget package?

**Some more info:**
On CoreCLR (DEBUG) you'd get a FileNotFoundException when you made the call into code that requires it. Unfortunately, the way .NET Native burns this information into the binary makes the Windows OS loader have a hard error on application start.

**Comment by MattWhilden** [comment]
Here's something else that noticed that I'll have to dig into tomorrow. You application currently uses Microsoft.NETCore.UniversalWindowsPlaftorm 5.2.2 and we're able to build you successfully. If you move to 5.3.0 of that package (only supported on Visual Studio 2017+) we seem to hit a snag during compilation. The error looks like this:

    MSBUILD : error : Error: NUTC300D:Internal Compiler Error: Type '$82_OpenCvSharp.Point3f[]' is not valid for use outside of the managed environment, but is passed to native code while computing call to 'static System.Void $82_OpenCvSharp.NativeMethods.calib3d_solvePnP_vector($82_OpenCvSharp.Point3f[], System.Int32, $82_OpenCvSharp.Point2f[], System.Int32, System.Double[,], System.Double[], System.Int32, System.Double[], System.Double[], System.Int32, System.Int32)'. while compiling method 'static System.Void $82_OpenCvSharp.Cv2.SolvePnP(System.Collections.Generic.IEnumerable`1<$82_OpenCvSharp.Point3f>, System.Collections.Generic.IEnumerable`1<$82_OpenCvSharp.Point2f>, System.Double[,], System.Collections.Generic.IEnumerable`1<System.Double>, System.Double[]&, System.Double[]&, System.Boolean, $82_OpenCvSharp.SolvePnPFlags)'.

**Comment by enghch** [comment]
Any chance the newer versions of Microsoft.NETCore.UniversalWindowsPlatform fixed this? There was a 5.3.1, 5.3.2, 5.3.3, 5.3.4 and now a 5.4.0

**Comment by MattWhilden** [comment]
I am able to build a project (DEBUG|RELEASE) (x86|x64) with these as references: 
```
    "Microsoft.NETCore.UniversalWindowsPlatform": "5.4.0",
    "OpenCvSharp3-AnyCPU": "3.2.0.20170419"
```

Do you have a project that's failing to build?
